## Model Fit in Linear Regression - Lab

## Introduction
In this lab, you'll learn how to evaluate your model results, and you'll learn methods to select the appropriate features using stepwise selection.

## Objectives
You will be able to:
* Analyze the results of regression and R-squared and adjusted-R-squared 
* Understand and apply forward and backward predictor selection

## The Boston Housing Data once more

We pre-processed the Boston Housing Data the same way we did before:

- We dropped "ZN" and "NOX" completely
- We categorized "RAD" in 3 bins and "TAX" in 4 bins
- We used min-max-scaling on "B", "CRIM" and "DIS" (and logtransformed all of them first, except "B")
- We used standardization on "AGE", "INDUS", "LSTAT" and "PTRATIO" (and logtransformed all of them first, except for "AGE") 

In [4]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
boston = load_boston()

boston_features = pd.DataFrame(boston.data, columns = boston.feature_names)
boston_features = boston_features.drop(["NOX","ZN"],axis=1)

# first, create bins for based on the values observed. 3 values will result in 2 bins
bins = [0,6,  24]
bins_rad = pd.cut(boston_features['RAD'], bins)
bins_rad = bins_rad.cat.as_unordered()

# first, create bins for based on the values observed. 4 values will result in 3 bins
bins = [0, 270, 360, 712]
bins_tax = pd.cut(boston_features['TAX'], bins)
bins_tax = bins_tax.cat.as_unordered()

tax_dummy = pd.get_dummies(bins_tax, prefix="TAX")
rad_dummy = pd.get_dummies(bins_rad, prefix="RAD")
boston_features = boston_features.drop(["RAD","TAX"], axis=1)
boston_features = pd.concat([boston_features, rad_dummy, tax_dummy], axis=1)

age = boston_features["AGE"]
b = boston_features["B"]
logcrim = np.log(boston_features["CRIM"])
logdis = np.log(boston_features["DIS"])
logindus = np.log(boston_features["INDUS"])
loglstat = np.log(boston_features["LSTAT"])
logptratio = np.log(boston_features["PTRATIO"])

# minmax scaling
boston_features["B"] = (b-min(b))/(max(b)-min(b))
boston_features["CRIM"] = (logcrim-min(logcrim))/(max(logcrim)-min(logcrim))
boston_features["DIS"] = (logdis-min(logdis))/(max(logdis)-min(logdis))

#standardization
boston_features["AGE"] = (age-np.mean(age))/np.sqrt(np.var(age))
boston_features["INDUS"] = (logindus-np.mean(logindus))/np.sqrt(np.var(logindus))
boston_features["LSTAT"] = (loglstat-np.mean(loglstat))/np.sqrt(np.var(loglstat))
boston_features["PTRATIO"] = (logptratio-np.mean(logptratio))/(np.sqrt(np.var(logptratio)))

In [3]:
boston_features

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0    0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1    0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2    0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3    0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4    0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   
5    0.162521 -1.778965   0.0  6.430 -0.351157  0.707895  0.165279  0.992990   
6    0.276046 -0.125137   0.0  6.012 -0.070229  0.671500 -1.496563  0.996722   
7    0.327656 -0.125137   0.0  6.172  0.978808  0.700059 -1.496563  1.000000   
8    0.367371 -0.125137   0.0  5.631  1.117494  0.709276 -1.496563  0.974104   
9    0.344658 -0.125137   0.0  6.004  0.616090  0.743201 -1.496563  0.974305   
10   0.373926 -0.125137   0.0  6.377  0.914799  0.727217 -1.496563  0.988956   
11   0.305940 -0.125137   0.0  6.009  0.509409  0.719175 -1.496563  1.000000   
12   0.282362 -0.125137   0.0  5.889 -1.051700  0.663113 -1.496563  0.983862   
13   0.481724 -0.081680   0.0  5.949 -0.240919  0.601338  1.095518  1.000000   
14   0.483078 -0.081680   0.0  6.096  0.566306  0.578763  1.095518  0.957436   
15   0.481329 -0.081680   0.0  5.834 -0.429390  0.582214  1.095518  0.996772   
16   0.535631 -0.081680   0.0  5.935 -1.396638  0.582214  1.095518  0.974658   
17   0.504684 -0.081680   0.0  5.990  0.466736  0.559046  1.095518  0.974406   
18   0.507126 -0.081680   0.0  5.456 -1.137046  0.510723  1.095518  0.727899   
19   0.496582 -0.081680   0.0  5.727  0.032897  0.510723  1.095518  0.984997   
20   0.553642 -0.081680   0.0  5.570  1.049929  0.510878  1.095518  0.948737   
21   0.513370 -0.081680   0.0  5.965  0.733440  0.534015  1.095518  0.988981   
22   0.552013 -0.081680   0.0  6.142  0.822342  0.530282  1.095518  1.000000   
23   0.528914 -0.081680   0.0  5.813  1.117494  0.542632  1.095518  0.994049   
24   0.500052 -0.081680   0.0  5.924  0.907687  0.572839  1.095518  0.993520   
25   0.511947 -0.081680   0.0  5.599  0.608978  0.578073  1.095518  0.764285   
26   0.488506 -0.081680   0.0  5.813  0.772557  0.599050  1.095518  0.949518   
27   0.525396 -0.081680   0.0  6.047  0.719216  0.577960  1.095518  0.771748   
28   0.503177 -0.081680   0.0  6.495  0.918355  0.578083  1.095518  0.977407   
29   0.530388 -0.081680   0.0  6.674  0.665875  0.557172  1.095518  0.957966   
..        ...       ...   ...    ...       ...       ...       ...       ...   
476  0.695890  0.947825   0.0  6.484  0.889907  0.300543  0.784047  0.998260   
477  0.813789  0.947825   0.0  5.304  1.021481  0.261386  0.784047  0.880428   
478  0.773591  0.947825   0.0  6.185  1.000145  0.275157  0.784047  0.956629   
479  0.808870  0.947825   0.0  6.229  0.690768  0.230282  0.784047  0.965757   
480  0.714587  0.947825   0.0  6.242 -0.137794  0.467238  0.784047  1.000000   
481  0.712484  0.947825   0.0  6.750  0.224924  0.455701  0.784047  0.990342   
482  0.712905  0.947825   0.0  7.061  0.299601  0.465562  0.784047  0.995915   
483  0.638603  0.947825   0.0  5.762 -1.005472  0.542950  0.784047  0.989964   
484  0.620842  0.947825   0.0  5.871 -0.948575  0.502600  0.784047  0.934011   
485  0.666348  0.947825   0.0  6.312 -0.592969  0.531847  0.784047  0.979121   
486  0.712182  0.947825   0.0  6.114  0.399171  0.481952  0.784047  0.989359   
487  0.695119  0.947825   0.0  5.905 -0.546740  0.432381  0.784047  0.978113   
488  0.332129  1.497881   0.0  5.454  0.857902  0.201163  0.744248  0.995436   
489  0.352559  1.497881   0.0  5.414  1.057042  0.185729  0.744248  0.866736   
490  0.365481  1.497881   0.0  5.093  1.046373  0.201557  0.744248  0.802133   
491  0.294927  1.497881   0.0  5.983  1.074822  0.211945  0.744248  0.982879   
492  0.300311  1.497881   0.0  5.983  0.530745  0.263227  0.744248  1.000000   
493  0.346652  0.142876   0.0  5.707 -0.518292  0.314279  0.376884  1.000000   
494  0.39

## Perform stepwise selection

The code for stepwise selection is copied below.

In [5]:
import statsmodels.api as sm

def stepwise_selection(X, y, 
                       initial_list=[], 
                       threshold_in=0.01, 
                       threshold_out = 0.05, 
                       verbose=True):
    """ Perform a forward-backward feature selection 
    based on p-value from statsmodels.api.OLS
    Arguments:
        X - pandas.DataFrame with candidate features
        y - list-like with the target
        initial_list - list of features to start with (column names of X)
        threshold_in - include a feature if its p-value < threshold_in
        threshold_out - exclude a feature if its p-value > threshold_out
        verbose - whether to print the sequence of inclusions and exclusions
    Returns: list of selected features 
    Always set threshold_in < threshold_out to avoid infinite looping.
    See https://en.wikipedia.org/wiki/Stepwise_regression for the details
    """
    included = list(initial_list)
    while True:
        changed=False
        # forward step
        excluded = list(set(X.columns)-set(included))
        new_pval = pd.Series(index=excluded)
        for new_column in excluded:
            model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit()
            new_pval[new_column] = model.pvalues[new_column]
        best_pval = new_pval.min()
        if best_pval < threshold_in:
            best_feature = new_pval.idxmin()
            included.append(best_feature)
            changed=True
            if verbose:
                print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

        # backward step
        model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
        # use all coefs except intercept
        pvalues = model.pvalues.iloc[1:]
        worst_pval = pvalues.max() # null if pvalues is empty
        if worst_pval > threshold_out:
            changed=True
            worst_feature = pvalues.argmax()
            included.remove(worst_feature)
            if verbose:
                print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
        if not changed:
            break
    return included

In [6]:
X = boston_features
y = pd.DataFrame(boston.target, columns= ["price"])

In [7]:
result = stepwise_selection(X, y, verbose = True)
print('resulting features:')
print(result)

Add  LSTAT                          with p-value 9.27989e-122
Add  RM                             with p-value 1.98621e-16
Add  PTRATIO                        with p-value 2.5977e-12
Add  DIS                            with p-value 2.85496e-09
Add  B                              with p-value 2.77572e-06
Add  TAX_(0, 270]                   with p-value 0.000855799
Add  CHAS                           with p-value 0.00151282
Add  INDUS                          with p-value 0.00588575
resulting features:
['LSTAT', 'RM', 'PTRATIO', 'DIS', 'B', 'TAX_(0, 270]', 'CHAS', 'INDUS']


### Build the final model again in Statsmodels

In [8]:
import statsmodels.api as sm
X_fin = X[["LSTAT", "RM", "PTRATIO", "DIS", "B", "TAX_(0, 270]", "CHAS", "INDUS"]]
X_int = sm.add_constant(X_fin)
model = sm.OLS(y,X_int).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.776
Model:                            OLS   Adj. R-squared:                  0.773
Method:                 Least Squares   F-statistic:                     215.7
Date:                Mon, 08 Apr 2019   Prob (F-statistic):          2.69e-156
Time:                        20:12:19   Log-Likelihood:                -1461.3
No. Observations:                 506   AIC:                             2941.
Df Residuals:                     497   BIC:                             2979.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            4.8980      2.813      1.742      0.082      -0.628      10.424
LSTAT           -5.5932      0.319    -17.538      0.000      -6.220      -4.967
RM               2.8294      0.386      7.333      0.000       2.071       3.587
PTRATIO         -1.3265      0.226     -5.878      0.000      -1.770      -0.883
DIS             -9.1984      1.333     -6.898      0.000     -11.818      -6.579
B                3.9052      0.931      4.195      0.000       2.076       5.734
TAX_(0, 270]     1.4418      0.552      2.614      0.009       0.358       2.526
CHAS             2.7988      0.791      3.539      0.000       1.245       4.353
INDUS           -0.9574      0.346     -2.766      0.006      -1.637      -0.277
==============================================================================
Omnibus:                      114.307   Durbin-Watson:                   1.088
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              482.579
Skew:                           0.945   Prob(JB):                    1.62e-105
Kurtosis:                       7.395   Cond. No.                         96.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Where our stepwise procedure mentions that "CHAS" was added with a p-value of 0.00151282, but our statsmodels output returns a p-value of 0.000. What is the intuition behind this?

## Use Feature ranking with recursive feature elimination

Use feature ranking to select the 5 most important features

In [9]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression

linreg = LinearRegression()
selector = RFE(linreg, n_features_to_select = 5)
selector = selector.fit(X, y)

/Users/blair/anaconda3/lib/python3.7/site-packages/sklearn/utils/validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [10]:
selector.support_ 

array([False, False,  True,  True, False,  True, False,  True,  True,
       False, False, False, False, False])

Fit the linear regression model again using the 5 columns selected

In [11]:
selected_columns = X.columns[selector.support_ ]
linreg.fit(X[selected_columns],y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

Now, predict $\hat y$ using your model. you can use `.predict()` in scikit-learn

In [12]:
yhat = linreg.predict(X[selected_columns])

In [13]:
yhat

array([[31.16165902],
       [23.76929453],
       [34.8764376 ],
       [36.98692436],
       [31.32930715],
       [29.06938762],
       [18.83320099],
       [14.70915944],
       [ 8.01751688],
       [14.90750061],
       [14.49961186],
       [17.84334784],
       [15.96671008],
       [23.35014491],
       [21.5475942 ],
       [22.80420786],
       [25.67823103],
       [17.66801955],
       [17.36837517],
       [19.86812537],
       [12.67589909],
       [18.42874828],
       [15.95977353],
       [14.09380242],
       [16.3420057 ],
       [13.99038842],
       [16.58165948],
       [15.0909717 ],
       [20.71269856],
       [22.06463506],
       [11.91943166],
       [19.11352766],
       [ 9.2779531 ],
       [14.33996426],
       [13.34927881],
       [22.57757018],
       [20.30738333],
       [22.88337432],
       [21.8018023 ],
       [31.92538034],
       [41.52527107],
       [31.06370984],
       [27.07153202],
       [24.77046631],
       [21.62254369],
       [20

Now, using the formulas of R-squared and adjusted-R-squared below, and your Python/numpy knowledge, compute them and contrast them with the R-squared and adjusted-R-squared in your statsmodels output using stepwise selection. Which of the two models would you prefer?

$SS_{residual} = \sum (y - \hat{y})^2 $

$SS_{total} = \sum (y - \bar{y})^2 $

$R^2 = 1- \dfrac{SS_{residual}}{SS_{total}}$

$R^2_{adj}= 1-(1-R^2)\dfrac{n-1}{n-p-1}$

In [14]:
import numpy as np
SS_Residual = np.sum((y-yhat)**2)
SS_Total = np.sum((y-np.mean(y))**2)
r_squared = 1 - (float(SS_Residual))/SS_Total
adjusted_r_squared = 1 - (1-r_squared)*(len(y)-1)/(len(y)-X[selected_columns].shape[1]-1)

In [15]:
r_squared

price    0.742981
dtype: float64

In [16]:
adjusted_r_squared

price    0.740411
dtype: float64

## Level up - Optional

- Perform variable selection using forward selection, using this resource: https://planspace.org/20150423-forward_selection_with_statsmodels/. Note that this time features are added based on the adjusted-R-squared!
- Tweak the code in the `stepwise_selection()`-function written above to just perform forward selection based on the p-value.

## Summary
Great! You now performed your own feature selection methods!